daemon/containerd/image_tag.go

```diff
@@ -1,20 +1,57 @@
 package containerd
 
 import (
-	"errors"
+	"context"
 
+	cerrdefs "github.com/containerd/containerd/errdefs"
+	containerdimages "github.com/containerd/containerd/images"
 	"github.com/docker/distribution/reference"
 	"github.com/docker/docker/errdefs"
 	"github.com/docker/docker/image"
+	"github.com/sirupsen/logrus"
 )
 
-// TagImage creates the tag specified by newTag, pointing to the image named
-// imageName (alternatively, imageName can also be an image ID).
-func (i *ImageService) TagImage(imageName, repository, tag string) (string, error) {
-	return "", errdefs.NotImplemented(errors.New("not implemented"))
-}
+// TagImageWithReference creates an image named as newTag and targeting the given descriptor id.
+func (i *ImageService) TagImageWithReference(ctx context.Context, imageID image.ID, newTag reference.Named) error {
+	logger := logrus.WithFields(logrus.Fields{
+		"imageID": imageID.String(),
+		"tag":     newTag.String(),
+	})
+
+	target, err := i.resolveDescriptor(ctx, imageID.String())
+	if err != nil {
+		logger.WithError(err).Debug("failed to resolve image id to a descriptor")
+		return err
+	}
+
+	img := containerdimages.Image{
+		Name:   newTag.String(),
+		Target: target,
+	}
+
+	is := i.client.ImageService()
+	_, err = is.Create(ctx, img)
+	if err != nil {
+		if !cerrdefs.IsAlreadyExists(err) {
+			logger.WithError(err).Debug("failed to create image")
+			return errdefs.System(err)
+		}
+
+		// If there already exists an image with this tag, delete it
+		err := is.Delete(ctx, img.Name)
```

Comment:
See [this](https://github.com/rumpl/moby/pull/127#discussion_r1070650316) comment, instead of deleting the existing image we should create a dangling image

Reply:
Done